core/src/test/java/io/parsingdata/metal/util/ReadTrackingByteStream.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.util;
+
+import static io.parsingdata.metal.Util.checkNotNull;
+import static java.util.stream.Collectors.toList;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.LongStream;
+
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ByteStream;
+
+public class ReadTrackingByteStream implements ByteStream {
+
+    private final ByteStream byteStream;
+    private final Set<Long> read = new HashSet<>();
```

Comment:
I really like that your implementation is very different from what I had in mind originally. Neat!